lib/autoupdate/rollout/strategy_timebased.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package rollout
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/gen/proto/go/teleport/autoupdate/v1"
+	update "github.com/gravitational/teleport/api/types/autoupdate"
+)
+
+const (
+	updateReasonInWindow      = "in_window"
+	updateReasonOutsideWindow = "outside_window"
+)
+
+type timeBasedStrategy struct {
+	log   *slog.Logger
+	clock clockwork.Clock
+}
+
+func (h *timeBasedStrategy) name() string {
+	return update.AgentsStrategyTimeBased
+}
+
+func newTimeBasedStrategy(log *slog.Logger, clock clockwork.Clock) (rolloutStrategy, error) {
+	if log == nil {
+		return nil, trace.BadParameter("missing log")
+	}
+	if clock == nil {
+		return nil, trace.BadParameter("missing clock")
+	}
+	return &timeBasedStrategy{
+		log:   log.With("strategy", update.AgentsStrategyTimeBased),
+		clock: clock,
+	}, nil
+}
+
+func (h *timeBasedStrategy) progressRollout(ctx context.Context, groups []*autoupdate.AutoUpdateAgentRolloutStatusGroup) error {
+	now := h.clock.Now()
+	// We always process every group regardless of the order.
+	var errors []error
```

Comment:
```suggestion
	var errs []error
```
nit: conflicts with `errors` package, so looks a bit confusing below